placement/algo/mirrored.go

```diff
@@ -205,12 +242,54 @@ func (a mirroredAlgorithm) ReplaceInstances(
 	return p, nil
 }
 
+// allInitializing returns true when the given list of instances
+// matches all the initializing instances in the placement.
+func allInitializing(instances []string, p placement.Placement) bool {
+	ids := make(map[string]struct{}, len(instances))
+	for _, i := range instances {
+		ids[i] = struct{}{}
+	}
+
+	for _, i := range p.Instances() {
+		if !i.IsInitializing() {
+			continue
+		}
+		if _, ok := ids[i.ID()]; !ok {
+			return false
+		}
+		delete(ids, i.ID())
+	}
+
+	return len(ids) == 0
+}
+
+// allInitializing returns true when the given list of instances
+// matches all the leaving instances in the placement.
+func allLeaving(instances []placement.Instance, p placement.Placement) bool {
+	ids := make(map[string]struct{}, len(instances))
+	for _, i := range instances {
+		ids[i.ID()] = struct{}{}
+	}
+
+	for _, i := range p.Instances() {
+		if !i.IsLeaving() {
```

Comment:
Can you refactor the common logic into a common `matchAllInState(...)` function and pass `func(i instance) bool { return i.IsLeaving() }` and  `func(i instance) bool { return i.IsInitializing() }` in these two functions? Reduces code duplication.

Reply:
sg